### PR TITLE
zigbee: Improve zboss trace support in ncp sample

### PR DESCRIPTION
**Pull request by sebastiandraus**

### CODE REVIEW

File: doc/nrf/libraries/zigbee/osif.rst [hunk 1]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+Received data is internally buffered and if Logging ZBOSS Traces is supported, data to transmit is also buffered.
+To configure this, use the following options:
+
+* :kconfig:`CONFIG_ZIGBEE_HAVE_ASYNC_SERIAL` - to enable Zigbee Async Serial.
+* :kconfig:`CONFIG_ZIGBEE_UART_DEVICE_NAME` - to specify serial device to use.
+* :kconfig:`CONFIG_ZIGBEE_UART_SUPPORTS_FLOW_CONTROL` - set this option for serial devices that support flow control.
+* :kconfig:`CONFIG_ZIGBEE_UART_RX_BUF_LEN` - to configure size of internal RX and TX buffer.
+* :kconfig:`ZBOSS_TRACE_BINARY_NCP_TRANSPORT_LOGGING` - to enable Logging ZBOSS Traces in binary format with Zigbee Async Serial.
+
+Backend options for Logging ZBOSS Traces:
```
Comment (greg-fer) [inline doc/nrf/libraries/zigbee/osif.rst:69]:
```suggestion
Backend functions for logging ZBOSS traces
===========================

The backend functions for logging ZBOSS traces belong to the following groups:

File: doc/nrf/libraries/zigbee/osif.rst [hunk 2]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+
+* Zigbee Async Serial - allows to use the same backend for NCP communication and Logging ZBOSS Traces. Logs ZBOSS Traces in binary format.
+  For how to configure, see :ref:`Zigbee Async Serial<zigbee_osif_zigbee_async_serial>`.
```
Comment (greg-fer) [inline doc/nrf/libraries/zigbee/osif.rst:72]:
```suggestion
* Zigbee Async Serial - This set of functions allows to use the same backend for NCP communication and logging ZBOSS traces.
When enabled, it logs ZBOSS traces in the binary format.
  For how to configure this backend, see `Zigbee async serial configuration options`_.
```
Althougn under the link provided there is barely any information about how to configure this option -- just one option is listed. If that's all, why not mention this option here? If more steps are needed, we need to expand that info.

File: doc/nrf/libraries/zigbee/osif.rst [hunk 3]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+
+* Zigbee Serial Logger - uses :ref:`Zephyr UART API<uart_api>` and can be configured to use UART peripheral or USB CDC ACM device.
+  Data is buffered internally in ring buffer and printed in binary format. Use the following options to configure it:
```
Comment (greg-fer) [inline doc/nrf/libraries/zigbee/osif.rst:75]:
```suggestion
* Zigbee serial logger - This set of functions uses Zephyr's :ref:`UART API <zephyr:uart_api>` and can be configured to use either the UART peripheral or the USB CDC ACM device.
  Data is buffered internally in ring buffer and printed in the binary format.
* Zigbee logger - This set of functions uses Zephyr's :ref:`zephyr:logging_api` API for logging hexdumps of received binary data.
  Data is buffered internally in ring buffer.

Zigbee serial logger configuration options
----------------------------------------

Use the following options to configure the Zigbee serial logger:

File: doc/nrf/libraries/zigbee/osif.rst [hunk 4]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+
+  * :kconfig:`CONFIG_ZBOSS_TRACE_BINARY_LOGGING` - to enable Logging ZBOSS Traces with Zigbee Serial Logger.
+  * :kconfig:`CONFIG_ZBOSS_TRACE_UART_LOGGING` - UART serial backend.
+  * :kconfig:`CONFIG_ZBOSS_TRACE_USB_CDC_LOGGING` - USB CDC ACM serial backend.
+  * :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_DEVICE_NAME` - to specify the serial device to use.
+  * :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_BUFFER_SIZE` - to specify size of internal ring buffer.
+
+* Zigbee logger - uses Zephyr's :ref:`zephyr:logging_api` API for Logging hexdumps of received binary data.
+  Data is buffered internally in ring buffer. Use the following options to configure it:
```
Comment (greg-fer) [inline doc/nrf/libraries/zigbee/osif.rst:84]: Zigbee logger configuration options
-------------------------------------

Use the following options to configure the Zigbee logger:

File: samples/zigbee/ncp/README.rst [hunk 5]
```diff
@@ -94,11 +94,27 @@ The USB device VID and PID are configured by the sample's Kconfig file.
 .. note::
     The USB is used as the default NCP communication channel when using the nRF52840 Dongle.
 
-You can configure :ref:`Zigbee stack logs <zigbee_ug_logging_stack_logs>` to be printed in binary format over the USB CDC using an independent serial port by setting:
+When communication channel is changed to nRF USB, :ref:`Zigbee stack logs <zigbee_ug_logging_stack_logs>` are being printed by default in binary format using ``UART1``.
+This is configured by setting:
 * :kconfig:`CONFIG_ZBOSS_TRACE_BINARY_LOGGING` - to enable binary format.
-* :kconfig:`CONFIG_ZBOSS_TRACE_USB_CDC_LOGGING` - to select USB CDC serial over UART serial.
+* :kconfig:`CONFIG_ZBOSS_TRACE_UART_LOGGING` - to select UART serial over nRF USB serial, set by default when binary format is enabled.
```
Comment (greg-fer) [inline samples/zigbee/ncp/README.rst:100]:
```suggestion
* :kconfig:`CONFIG_ZBOSS_TRACE_UART_LOGGING` - to select the UART serial over the nRF USB serial.
  This option is set by default when the binary format is enabled.
```

File: samples/zigbee/ncp/README.rst [hunk 6]
```diff
@@ -94,11 +94,27 @@ The USB device VID and PID are configured by the sample's Kconfig file.
...
+* :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_DEVICE_NAME` - to select the serial device to use for printing Zigbee stack logs, set to ``"UART_1"`` when UART serial selected.
```
Comment (greg-fer) [inline samples/zigbee/ncp/README.rst:101]:
```suggestion
* :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_DEVICE_NAME` - to select the serial device to use for printing Zigbee stack logs.
   This option is set to ``"UART_1"`` when the UART serial selected.
```

File: samples/zigbee/ncp/README.rst [hunk 7]
```diff
@@ -94,11 +94,27 @@ The USB device VID and PID are configured by the sample's Kconfig file.
...
 
-No additional changes in application are required. For more configuration options, see :ref:`Zigbee stack logs <zigbee_ug_logging_stack_logs>`.
```
Comment (greg-fer) [inline samples/zigbee/ncp/README.rst:101]: Add blank line after a list.

File: samples/zigbee/ncp/README.rst [hunk 8]
```diff
@@ -94,11 +94,27 @@ The USB device VID and PID are configured by the sample's Kconfig file.
...
+* :kconfig:`CONFIG_ZBOSS_TRACE_BINARY_LOGGING` - to enable binary format.
+* :kconfig:`CONFIG_ZBOSS_TRACE_USB_CDC_LOGGING` - to select nRF USB serial over UART serial.
+* :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_DEVICE_NAME` - to select the serial device for printing Zigbee stack logs, set to ``"CDC_ACM_1"``.
+
+To create two instances of USB CDC ACM for application, set the following:
+* :kconfig:`CONFIG_USB_CDC_ACM_DEVICE_COUNT` - defines number of USB CDC ACM devices to create, set to ``2`` in this configuration of NCP sample.
+* :kconfig:`CONFIG_USB_COMPOSITE_DEVICE` - to enable composite USB device driver.
+
+.. note::
+    With this configuration, you will see two serial ports created by the NCP sample. use the first one for NCP communication and second one to collect Zigbee stack logs.
```
Comment (greg-fer) [inline samples/zigbee/ncp/README.rst:114]: Split sentences to new lines and capitalize.

**Review by greg-fer — CHANGES_REQUESTED** [review]
Apologies for providing changes in this form, but I'm having issues with building locally and was jumping back and forth between tasks.

File: doc/nrf/libraries/zigbee/osif.rst [hunk 9]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
 * :kconfig:`CONFIG_ZBOSS_OSIF_LOG_LEVEL` - Configures the custom logger options for the ZBOSS OSIF layer.
 
+.. _zigbee_osif_zboss_osif_serial:
+
+ZBOSS OSIF Serial
+*****************
+
+Setting the :kconfig:`CONFIG_ZIGBEE_HAVE_SERIAL` enables serial abstract for the ZBOSS OSIF.
+
+ZBOSS OSIF serial implements a set of functions used by ZBOSS stack for serial communication, from which two sets of functions with common purpose can be distinguish:
```
[the remark this answers is not in the document]
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:51]: We implements multiple functions, some of them are used for NCP communication, others are for logging traces.

File: doc/nrf/libraries/zigbee/osif.rst [hunk 10]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+
+* Handling NCP communication with HOST device - used only for :ref:`NCP architecture <ug_zigbee_platform_design_ncp>`
+* Logging ZBOSS Traces - stack logs useful for debugging, received in binary format.
+
+.. _zigbee_osif_zigbee_async_serial:
```
[the remark this answers is not in the document]
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:56]: No header for this - internal use only

File: doc/nrf/libraries/zigbee/osif.rst [hunk 11]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+
+ZBOSS OSIF serial supports single backend for handling NCP communication - Zigbee Async Serial.
+It uses :ref:`Zephyr UART API<uart_api>` and can be configured to use UART peripheral or USB CDC ACM device.
```
[the remark this answers is not in the document]
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:59]: Using Kconfig option `CONFIG_ZIGBEE_UART_DEVICE_NAME` to select USB Serial.
This requires USB config to be enabled, NCP sample has it configured when built with `prj_usb.conf` conf file.

File: doc/nrf/libraries/zigbee/osif.rst [hunk 12]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+Received data is internally buffered and if Logging ZBOSS Traces is supported, data to transmit is also buffered.
```
[the remark this answers is not in the document]
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:60]: There are three "backends" to choose from for logging ZBOSS traces and only one can be chosen.
To use Zigbee Async Serial for logging traces, we enable `ZBOSS_TRACE_BINARY_NCP_TRANSPORT_LOGGING`.

File: doc/nrf/libraries/zigbee/osif.rst [hunk 13]
```diff
@@ -40,6 +41,55 @@ Additionally, the following Kconfig option is available when setting :ref:`zigbe
...
+To configure this, use the following options:
+
+* :kconfig:`CONFIG_ZIGBEE_HAVE_ASYNC_SERIAL` - to enable Zigbee Async Serial.
+* :kconfig:`CONFIG_ZIGBEE_UART_DEVICE_NAME` - to specify serial device to use.
+* :kconfig:`CONFIG_ZIGBEE_UART_SUPPORTS_FLOW_CONTROL` - set this option for serial devices that support flow control.
+* :kconfig:`CONFIG_ZIGBEE_UART_RX_BUF_LEN` - to configure size of internal RX and TX buffer.
+* :kconfig:`ZBOSS_TRACE_BINARY_NCP_TRANSPORT_LOGGING` - to enable Logging ZBOSS Traces in binary format with Zigbee Async Serial.
```
[the remark this answers is not in the document]
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:67]: Should be, will do

[the remark this answers is not in the document]
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:67]: will do

File: doc/nrf/libraries/zigbee/osif.rst (hunk 2 above)
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:72]: Will check this

File: doc/nrf/libraries/zigbee/osif.rst (hunk 4 above)
Comment (sebastiandraus) [reply doc/nrf/libraries/zigbee/osif.rst:84]: will do

File: samples/zigbee/ncp/README.rst (hunk 7 above)
Comment (sebastiandraus) [reply samples/zigbee/ncp/README.rst:101]: will do

File: samples/zigbee/ncp/README.rst (hunk 8 above)
Comment (sebastiandraus) [reply samples/zigbee/ncp/README.rst:114]: will do

File: samples/zigbee/ncp/README.rst [hunk 14]
```diff
@@ -94,11 +94,32 @@ The USB device VID and PID are configured by the sample's Kconfig file.
...
+This is configured by setting:
+
+* :kconfig:`CONFIG_ZBOSS_TRACE_BINARY_LOGGING` - to enable binary format.
+* :kconfig:`CONFIG_ZBOSS_TRACE_UART_LOGGING` - to select the UART serial over the nRF USB serial.
+  This option is set by default when the binary format is enabled.
+* :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_DEVICE_NAME` - to select the serial device to use for printing Zigbee stack logs.
+  This option is set to ``"UART_1"`` when the UART serial selected.
+
+Alternatively, you can configure :ref:`Zigbee stack logs <zigbee_ug_logging_stack_logs>` to be printed in binary format using independent CDC ACM device of the same nRF USB.
+This can be configured by setting:
```
Comment (sebastiandraus) [inline samples/zigbee/ncp/README.rst:107]: This list is not displayed correctly

File: samples/zigbee/ncp/README.rst [hunk 15]
```diff
@@ -94,11 +94,32 @@ The USB device VID and PID are configured by the sample's Kconfig file.
...
 * :kconfig:`CONFIG_ZBOSS_TRACE_BINARY_LOGGING` - to enable binary format.
-* :kconfig:`CONFIG_ZBOSS_TRACE_USB_CDC_LOGGING` - to select USB CDC serial over UART serial.
+* :kconfig:`CONFIG_ZBOSS_TRACE_USB_CDC_LOGGING` - to select nRF USB serial over UART serial.
+* :kconfig:`CONFIG_ZBOSS_TRACE_LOGGER_DEVICE_NAME` - to select the serial device for printing Zigbee stack logs, set to ``"CDC_ACM_1"``.
+
+To create two instances of USB CDC ACM for application, set the following:
```
Comment (sebastiandraus) [inline samples/zigbee/ncp/README.rst:112]: This list is not displayed correctly

**Review by greg-fer — APPROVED** [review]

**Comment by greg-fer** [comment]
Please remember to squash commits before merge!

**Comment by sebastiandraus** [comment]
Commits squashed

**Review by totyz — APPROVED** [review]

**Pull request merged**